### ISSUE TITLE
NixOS fails to mount partitions with "label=partitionlabel" because it is wrongly using by-label instead of by-partlabel

### ISSUE THREAD

**Issue by ahydronous** [NixOS/nixpkgs#271731]
### Describe the bug
NixOS can't find partitions via label because it tries to use the paths `/dev/disk/by-label/mypartition` whereas NixOS mounts via `/dev/disk/by-partlabel/partition`

### Steps To Reproduce

Use `swapDevices = [ { label = "myswap"; } ];` with "myswap" being the label set on on the swap partition

System then complains about "looking for /dev/disk/by-label/myswap"

### Expected behavior
System mounts the swap partition and boots without error

### Fix
`label =` should use `by-partlabel`, not `by-label`.

**Comment by eclairevoyant** [comment]
Strongly disagree. You cannot mount a partition (partlabel), only a filesystem (label). So the natural meaning of `.label` should be the FS label. Your suggestion is the far more surprising behaviour.

If you want to use partlabel, you'd have to extend the `fileSystems` submodule.

**Comment by ahydronous** [comment]
> You cannot mount a partition (partlabel)

I mean, that's literally what I'm doing to workaround this bug, via `device = /dev/disk/by-partlabel/mypartition`

Perhaps the name of `.label` should be changed to something less ambiguous.

**Comment by ahydronous** [comment]
. .label means label, they're quite literally the same word, I fail to see the ambiguity. If you'd like .partlabel perhaps a PR is in order?

... you cannot see how `by-label` and `by-partlabel` are horribly ambiguous?
Not to mention that `by-label` is wrong, its called partition **name**, precisely to prevent this ambiguity and confusion.

**Comment by eclairevoyant** [comment]
> calling `by-label` a label is wrong these days, its called partition **name**

I think you're confused, `by-label` is not a property of a partition, so "partition name" would be horribly wrong. https://www.mankier.com/8/mke2fs calls it a "volume label".

Anyway, think I've made my original point clearly enough, if others feel differently I leave it up to them.

**Comment by ahydronous** [comment]
I am aware that `by-label` is a property of GPT whereas `by-partlabel` lives in the filesystem.
The NixOS manual itself suggests "root-partition" as an example label.

Let's just take Gparted as the lead here, I trust they have thought this through much more deeply than either of us :)

`by-label` is what corresponds to `Name` in Gparted, and `by-partlabel` corresponds to `Label` in Gparted.
NixOS is flouting this convention which leads to users making assumptions and getting into trouble.

Note that `fileSystems.<name>.encrypted.label` also use the `label` terminology.

Perhaps `fileSystems.label` should be renamed `fileSystems.gptlabel` and `partlabel` should become `fileSystems.fslabel` to prevent any future confusion.

**Comment by yl3dy** [comment]
> 1. `.label` is too ambiguous to someone glancing over it
>        For me it was lucky because it was just my swap partition, but I imagine someone in the future will also assume `.label` refers to `by-partlabel` (which I feel is not a gigantic leap to make), footgun themselves and lose access to their root partition. Then they'll either need to be comfortable in SystemD's emergency shell, or have a LiveUSB at the ready.

True, these names can be confusing. However, in NixOS it's not that much of a problem: you can just select the previous generation in the bootloader (just with arrow keys in systemd-boot, a separate menu in GRUB IIRC) and boot to the last good configuration. Also, in the case of swap, you can do a `nixos-rebuild switch` which would just show an error without crashing the system if your labels are wrong.


> 2\. `.partlabel` doesn't exist yet
>         Like mentioned by @eclairevoyant, this needs to be written. I might check out how `.label` works and take it from there

It doesn't seem like a very popular feature, but don't hesitate to try to implement it. You'll get acquianted with Nixpkgs module system in the process! But IMO it should be a separate option: it won't do good to guess too much here.


> 3\. `encrypted.label` also uses `.label`
>         This is much less of an issue / priority as here the label refers to `/dev/mapper/<label>` and as I understand (wrongly?) that's the only way to really use a label with an encrypted device, but I feel it should be included for posterity as we're already talking about `.label` being aspecific

Well, _this_ `label` is really confusing, because it doesn't have anything in common with filesystem labels :smile: It's connected with how device-mapper works. It's the transparent block layer between a physical device (or devices) and some useful representation as a block device inside of `/dev/mapper/` with arbitrary name ("mapped name"), and that device can be mounted as a filesystem. Two common applications are LVM (Logical Volume Manager, can be used to merge several physical drives into a single physical partition, `/dev/mapper/<volume-group-name>-<volume-name>` is a block device of a specific volume) and dm-crypt (usually used with LUKS, the block device in `/dev/mapper` is the unencrypted partition). So in case of LUKS, the manual procedure of mounting might look like this: unlock the encrypted partition with `cryptsetup luksOpen /dev/sda1 nix_rocks` and mount it with `mount /dev/mapper/nix_rocks /mnt/data`. Note that the mapped name (in this case `nix_rocks`) can be replaced with anything you like, and _that's_ the label in `swapDevices.*.encrypted.label` and `fileSystems.<name>.encrypted.label`.

However, do note that in the case of LUKS-encrypted root, you'd like to do it a bit differently. First, specify it to be unlocked in initrd via `boot.initrd.luks.devices."nvme0n1p6_crypt".device = "/dev/disk/by-uuid/<uuid>";` (or using labels). Then you can mount it using its mapped name (`nvme0n1p6` here) via `fileSystems."/" = { device = "/dev/mapper/nvme0n1p6_crypt"; ... };` (you can use the unencrypted filesystem's UUID here too, that's what `nixos-generate-config` does). This way you'd have all necessary pieces ready at appropriate times during boot.

**Comment by bugQ** [comment]
If I may offer a opinion:

> ... you cannot see how `by-label` and `by-partlabel` are horribly ambiguous? Not to mention that calling `by-label` a label is wrong these days, its called partition **name**, precisely to prevent this ambiguity and confusion.
> 1. `.label` is too ambiguous to someone glancing over it
>    For me it was lucky because it was just my swap partition, but I imagine someone in the future will also assume `.label` refers to `by-partlabel` (which I feel is not a gigantic leap to make), footgun themselves and lose access to their root partition. Then they'll either need to be comfortable in SystemD's emergency shell, or have a LiveUSB at the ready.

It is clear that you experienced confusion, and hopefully it is clear by now that `.label` does not (and should not) refer to `by-partlabel`. It's worth noting that this terminology comes directly from the Linux project, specifically `util-linux`, which includes tools such `lsblk` that refer to filesystem labels with `LABEL` and partition labels with `PARTLABEL`, notwithstanding that some tools such as KDE Partition Manager refer to the latter as "partition name". (In fact, it may have only added to the confusion.) If that is indeed the new naming convention "these days", then somebody ought to tell the `util-linux` team.

That being said, adding `.partlabel` as an option alongside `.label` may make it more clear to NixOS users  what the differences, and now that #296438 is open, I suggest that this issue be closed.

**Comment by eclairevoyant** [comment]
It'd make more sense to close this only once the PR is closed.
